commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/MongoRef.scala

```diff
@@ -0,0 +1,329 @@
+package com.avsystem.commons
+package mongo.typed
+
+import com.avsystem.commons.annotation.macroPrivate
+import com.avsystem.commons.meta.OptionLike
+import com.avsystem.commons.mongo.typed.MongoPropertyRef.Separator
+import com.avsystem.commons.mongo.{BsonValueInput, KeyEscaper}
+import com.avsystem.commons.serialization.GenCodec.ReadFailure
+import org.bson.{BsonDocument, BsonValue}
+
+/**
+  * Represents a reference to a particular "place" in a MongoDB document. The "place" may be an actual path
+  * inside the document ([[MongoPropertyRef]]) or the whole document _itself_ (you can think of it as an empty path).
+  *
+  * When the [[MongoRef]] points to the whole document, it may also narrow the type only to some subtype(s).
+  * See the [[DataRefDsl.as]] macro for more details on narrowing.
+  *
+  * @tparam E the data type
+  */
+sealed trait MongoRef[E, T] extends MongoProjection[E, T] with DataRefDsl[E, T] { self =>
+  def format: MongoFormat[T]
+  def projectionRefs: Set[MongoRef[E, _]] = Set(this)
+  def showRecordId: Boolean = false
+
```

Comment:
This gets a bit complicated because `MongoRef` is technically a _partial_ function now. Also, for consistency I would probably have to add this on `MongoProjection` level and maybe `MongoFilter` should also have a memory-level filter, etc...